test/falco_k8s_audit_tests.yaml

```diff
@@ -128,6 +128,16 @@ trace_files: !mux
       - Create Privileged Pod: 1
     trace_file: trace_files/k8s_audit/create_nginx_pod_privileged.json
 
+  create_privileged_no_secctx_1st_container_2nd_container_pod:
```

Comment:
I don't understand how this test is related to the NA value in the code. Can you help me undertstand?

Reply:
This test ensures that for a k8s audit event where the first container doesn't have any security context and the 2nd container does have a security context and privileged is true, you still get a falco alert for the Create Privileged Pod rule.

Behind the scenes, what's changing is that older falco versions used to abort when trying to get the value for the ka.req.pod.containers.privileged field. Whenever a field can't be extracted from an event, the result is "<NA>". With this fix, what's returned instead is an array with values for each container, where each container will have "true" if the security context is present and true, "false" if present and false, and "<NA>" otherwise. Here's an example minimal rules file that demonstrates the difference:

```
- list: k8s_audit_stages
  items: ["ResponseComplete"]

- macro: kevt
  condition: (jevt.value[/stage] in (k8s_audit_stages))

- macro: kcreate
  condition: ka.verb=create

- macro: pod
  condition: ka.target.resource=pods and not ka.target.subresource exists

- rule: Create Pod
  desc: Detect any attempt to create a pod
  condition: kevt and pod and kcreate
  output: Pod created (user=%ka.user.name pod=%ka.resp.name ns=%ka.target.namespace images=%ka.req.pod.containers.image privileged=%ka.req.pod.containers.privileged)
  priority: INFO
  source: k8s_audit
  tags: [k8s]
  ```
  
  ```
$  sudo falco -r ~/create_pod.yaml -e create_nginx_pod_no_secctx_1st_container_privileged_2nd_container.json
Wed Apr  7 10:13:33 2021: Falco version 0.27.0-84+8c9d4f4 (driver version 5c0b863ddade7a45568c0ac97d037422c9efb750)
Wed Apr  7 10:13:33 2021: Falco initialized with configuration file /etc/falco/falco.yaml
Wed Apr  7 10:13:33 2021: Loading rules from file /home/mstemm/create_pod.yaml:
Wed Apr  7 10:13:33 2021: Reading k8s audit events from file: /mnt/sf_mstemm/work/src/falco/test/trace_files/k8s_audit/create_nginx_pod_no_secctx_1st_container_privileged_2nd_container.json
Events detected: 1
Rule counts by severity:
   INFO: 1
Triggered rules by rule name:
   Create Pod: 1
Syscall event drop monitoring:
   - event drop detected: 0 occurrences
   - num times actions taken: 0
 10:53:07.006844928: Notice Pod created (user=system:serviceaccount:kube-system:replicaset-controller pod=nginx-deployment-544b59f8b8-ffkxm ns=default images=(nginx,mysql:latest) privileged=<NA>)

 $ sudo ./userspace/falco/falco -r ~/create_pod.yaml -e create_nginx_pod_no_secctx_1st_container_privileged_2nd_container.json
Wed Apr  7 10:14:09 2021: Falco version 0.27.0-84+8c9d4f4 (driver version 5c0b863ddade7a45568c0ac97d037422c9efb750)
Wed Apr  7 10:14:09 2021: Falco initialized with configuration file /etc/falco/falco.yaml
Wed Apr  7 10:14:09 2021: Loading rules from file /home/mstemm/create_pod.yaml:
Wed Apr  7 10:14:09 2021: Reading k8s audit events from file: /mnt/sf_mstemm/work/src/falco/test/trace_files/k8s_audit/create_nginx_pod_no_secctx_1st_container_privileged_2nd_container.json
Events detected: 1
Rule counts by severity:
   INFO: 1
Triggered rules by rule name:
   Create Pod: 1
Syscall event drop monitoring:
   - event drop detected: 0 occurrences
   - num times actions taken: 0
10:53:07.006844928: Notice Pod created (user=system:serviceaccount:kube-system:replicaset-controller pod=nginx-deployment-544b59f8b8-ffkxm ns=default images=(nginx,mysql:latest) privileged=(<NA>,true))
```

Reply:
Thanks for the exhaustive explanation Mark!